invertedai/large/drive.py

```diff
@@ -0,0 +1,314 @@
+        currently supports 'car' and 'pedestrian'.
+        waypoint: optional [Point], the target waypoint of the agent.
+
+    recurrent_states:
+        Recurrent states for all agents, obtained from the previous call to
+        :func:`drive` or :func:`initialize`.
+
+    traffic_lights_states:
+       If the location contains traffic lights within the supported area,
+       their current state should be provided here. Any traffic light for which no
+       state is provided will have a state generated by iai.
+
+    light_recurrent_states:
+       Light recurrent states for all agents, obtained from the previous call to
+        :func:`drive` or :func:`initialize`.
+       Specifies the state and time remaining for each light group in the map.
+       If manual control of individual traffic lights is desired, modify the relevant state(s) 
+       in traffic_lights_states, then pass in light_recurrent_states as usual.
+
+    get_birdview:
+        Whether to return an image visualizing the simulation state.
+        This is very slow and should only be used for debugging.
+
+    rendering_center:
+        Optional center coordinates for the rendered birdview.
+
+    rendering_fov:
+        Optional fov for the rendered birdview.
+
+    get_infractions:
+        Whether to check predicted agent states for infractions.
+        This introduces some overhead, but it should be relatively small.
+
+    random_seed:
+        Controls the stochastic aspects of agent behavior for reproducibility.
+
+    api_model_version:
+        Optionally specify the version of the model. If None is passed which is by default, 
+        the best model will be used.
+    
+    capacity:
+        The number of agents allowed in a region before it must subdivide. Currently this 
+        value represents the capacity of a quadtree leaf node that will subdivide if the 
+        number of vehicles in the region passes this threshold.
+
+    is_async:
+        A flag to control whether to use asynchronous DRIVE calls.
+
+    See Also
+    --------
+    :func:`drive`
+    """
+    
+    num_agents = len(agent_states)
+    agent_x = [None]*num_agents
+    agent_y = [None]*num_agents
+    for i, agent in enumerate(agent_states):
+        agent_x[i] = agent.center.x
+        agent_y[i] = agent.center.y
+    max_x, min_x, max_y, min_y = max(agent_x), min(agent_x), max(agent_y), min(agent_y)
+    region_size = ceil(max(max_x,max_y) - min(min_x,min_y))
+    region_center = ((max_x+min_x)/2,(max_y+min_y)/2)
+
+    quadtree = QuadTree(
+        capacity=capacity,
+        region=Region.init_square_region(
+            Point.fromlist(list(region_center)),
+            region_size
+        ),
+    )
+    for i, (agent, attrs, recurr_state) in enumerate(zip(agent_states,agent_attributes,recurrent_states)):
+        agent_info = AgentInfo.fromlist([agent, attrs, recurr_state, i])
+        is_inserted = quadtree.insert(agent_info)
```

Comment:
It would only ever be false if the agent was outside of the regions defined by the quad tree which would be an error so I've added an appropriate error message.